src/app/util/attribute-storage.cpp

```diff
@@ -266,9 +266,64 @@ uint16_t emberAfGetDynamicIndexFromEndpoint(EndpointId id)
     return kEmberInvalidEndpointIndex;
 }
 
+namespace {
+
+const EmberAfCluster * getClusterTypeDefinition(EndpointId endpointId, ClusterId clusterId, EmberAfClusterMask mask)
+{
+    uint16_t index = emberAfIndexFromEndpointIncludingDisabledEndpoints(endpointId);
+    if (index != kEmberInvalidEndpointIndex)
+    {
+        return emberAfFindClusterInType(emAfEndpoints[index].endpointType, clusterId, mask, nullptr);
+    }
+    // not found
+    return nullptr;
+}
+
+} // anonymous namespace
+
+void emberAfSetupDynamicEndpointDeclaration(EmberAfEndpointType & endpointType, EndpointId templateEndpointId,
+                                            const Span<const EmberAfClusterSpec> & templateClusterSpecs)
+{
```

Comment:
Todo on this method:

- [ ] assert on endpointType.cluster being null
- [ ] create the `EmberAfCluster[]` array in a local variable so you can modify it (no const cast) and assign it at the end
- [ ] use chip::CanCastTo to validate that the clusterCount can receive all templateClusters
- [ ] make this method return a status: you can fail on too many clusters, on failure to allocate memory. This is better than having chipDie

Reply:
Finally, I think I understood and fixed all the todos ;-) Thanks for your help.